File: middleware.ts
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'

// ─── In-memory rate limiter (local dev only) ────────────────────────────────
//
// ⚠️  PRODUCTION SWAP: Replace this Map with Upstash Redis or Vercel KV.
//
//   import { Redis } from '@upstash/redis'
//   const redis = new Redis({ url: process.env.UPSTASH_REDIS_REST_URL!,
//                             token: process.env.UPSTASH_REDIS_REST_TOKEN! })
//
//   Then inside the middleware, use:
//     const key = `ratelimit:${ip}`
//     const count = await redis.incr(key)
//     if (count === 1) await redis.expire(key, 10)   // 10-second window
//     if (count > 5) return new NextResponse(...)
//
// ─────────────────────────────────────────────────────────────────────────────

const rateLimitMap = new Map<string, { count: number; resetAt: number }>()

const WINDOW_MS = 10_000   // 10 seconds
const MAX_REQUESTS = 5     // 5 requests per window

export function middleware(request: NextRequest) {
  const ip = request.headers.get('x-forwarded-for') ??
             request.headers.get('x-real-ip') ??
             '127.0.0.1'

  const now = Date.now()
  const record = rateLimitMap.get(ip)

  if (!record || now > record.resetAt) {
    rateLimitMap.set(ip, { count: 1, resetAt: now + WINDOW_MS })
    return NextResponse.next()
  }

  record.count += 1

  if (record.count > MAX_REQUESTS) {
    return NextResponse.json(
      { error: 'Too Many Requests', retryAfter: Math.ceil((record.resetAt - now) / 1000) },
      { status: 429, headers: { 'Retry-After': String(Math.ceil((record.resetAt - now) / 1000)) } }
    )
  }

  return NextResponse.next()
}

// Apply middleware only to API routes
export const config = {
  matcher: '/api/:path*',
}

File: backend/middleware.py
import time
from collections import defaultdict
from fastapi import Request, Response
from starlette.middleware.base import BaseHTTPMiddleware

# ─── In-memory rate limiter (local dev only) ─────────────────────────────────
#
# ⚠️  PRODUCTION SWAP: Replace this dict with Redis / Upstash KV.
#
#   import redis
#   r = redis.from_url(os.getenv("REDIS_URL"))
#
#   Then inside the middleware:
#     key = f"ratelimit:{client_ip}"
#     count = r.incr(key)
#     if count == 1:
#         r.expire(key, WINDOW_SECONDS)
#     if count > MAX_REQUESTS:
#         return JSONResponse(status_code=429, content={...})
#
# ─────────────────────────────────────────────────────────────────────────────

WINDOW_SECONDS = 10
MAX_REQUESTS = 5

# {ip: [(timestamp, ...), ...]}
rate_limit_store: dict[str, list[float]] = defaultdict(list)


class RateLimitMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next):
        client_ip = request.client.host if request.client else "unknown"
        now = time.monotonic()

        # Prune expired entries outside the window
        rate_limit_store[client_ip] = [
            t for t in rate_limit_store[client_ip]
            if now - t < WINDOW_SECONDS
        ]

        if len(rate_limit_store[client_ip]) >= MAX_REQUESTS:
            return Response(
                status_code=429,
                media_type="application/json",
                content='{"error":"Too Many Requests","retry_after":' + str(WINDOW_SECONDS) + '}',
                headers={"Retry-After": str(WINDOW_SECONDS)},
            )

        rate_limit_store[client_ip].append(now)
        response = await call_next(request)
        return response
